lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,388 @@
+        curl_args = (
+            base_curl_fetch_args(
+                self.url, self.timeout, status_bar=False, user_agent=SPACK_USER_AGENT
+            )
+            + self._config_args
+        )
+        curl_cmd = [self._curl] + curl_args
+        self._curl_process = subprocess.Popen(
+            curl_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+
+        self._scheme = urllib.parse.urlparse(self.url).scheme
+        assert self._curl_process.stdout is not None, "curl process stderr is None"
+        self._stream = io.BufferedReader(self._curl_process.stdout)  # type: ignore[type-var]
+
+        # curl echoes intermediate redirect responses, so we might get multiple headers
+        headers = []
+        while True:
+            header, finished = self._get_next_header()
+            headers.append(header)
+            if finished:
+                break
+
+        if not headers:
+            raise FetchError(f"Failed to fetch {self.url}: no headers returned")
+
+        self._headers = headers
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._curl_process.wait()
+        check_curl_code(self._curl_process.returncode)
+
+    def read(self, size: int) -> bytes:
+        return self._stream.read(size)  # type: ignore
+
+    def request_info(self) -> RequestInfo:
+        effective_url = self.url
+        for h in reversed(self._headers):
+            if "location" in h:
+                effective_url = h["location"]
+                break
+        return RequestInfo(url=self.url, effective_url=effective_url, headers=self._headers[-1])
+
+    @staticmethod
+    def cookie_args(cookie) -> List[str]:
+        """Returns the arguments to pass to curl to use a cookie.
+
+        This can be used as a helper function to generate the arguments for the constructor.
+        """
+        return ["-j", "-b", cookie]
+
+    def _get_next_header(self):
+        """Returns the next header from the stream."""
+
+        if self._scheme not in ("http", "https"):
+            return http.client.HTTPMessage(), True
+
+        finished = True
+        status_line = self._stream.readline().decode("iso-8859-1")
+        if not status_line.startswith("HTTP/"):
+            raise FetchError(f"Failed to fetch {self.url}: unexpected status line: {status_line}")
+        try:
+            status = int(status_line.split()[1])
+        except ValueError:
+            raise FetchError(
+                f"Failed to fetch {self.url}: cannot parse HTTP status code from {status_line}"
+            )
+
+        if 400 <= status < 600:
+            header = http.client.parse_headers(self._stream)
+            raise DetailedHTTPError(
+                urllib.request.Request(self.url),
+                status,
+                HTTPStatus(int(status)).phrase,
+                header,
+                None,
+            )
+
+        elif 300 <= status < 400:
+            finished = False
```

Comment:
finished = 300 <= status < 400

Reply:
Hmm, the fact that this does the opposite maybe suggests we should leave it 2 lines?

Reply:
That's a good point